src/aks-preview/azext_aks_preview/decorator.py

```diff
@@ -863,6 +871,201 @@ def get_gmsa_dns_server_and_root_domain_name(self) -> Tuple[Union[str, None], Un
+        """Obtain the value of kubernetes_version.
+
+        Note: Inherited and extended in aks-preview to add support for getting values from snapshot.
+
+        :return: string
+        """
+        return self._get_kubernetes_version()
+
+    # pylint: disable=unused-argument
+    def _get_os_sku(self, read_only: bool = False, **kwargs) -> Union[str, None]:
+        """Internal function to dynamically obtain the value of os_sku according to the context.
+
+        If snapshot_id is specified, dynamic completion will be triggerd, and will try to get the corresponding value
+        from the Snapshot. When determining the value of the parameter, obtaining from `mc` takes precedence over user's
+        explicit input over snapshot over default vaule.
+
+        :return: string or None
+        """
+        # read the original value passed by the command
+        raw_value = self.raw_param.get("os_sku")
+        # try to read the property value corresponding to the parameter from the `mc` object
+        value_obtained_from_mc = None
+        if self.mc and self.mc.agent_pool_profiles:
+            agent_pool_profile = safe_list_get(
+                self.mc.agent_pool_profiles, 0, None
+            )
+            if agent_pool_profile:
+                value_obtained_from_mc = agent_pool_profile.os_sku
+        # try to retrieve the value from snapshot
+        value_obtained_from_snapshot = None
+        # skip dynamic completion if read_only is specified
+        if not read_only:
+            snapshot = self.get_snapshot()
+            if snapshot:
+                value_obtained_from_snapshot = snapshot.os_sku
+
+        # set default value
+        if value_obtained_from_mc is not None:
+            os_sku = value_obtained_from_mc
+        elif raw_value is not None:
+            os_sku = raw_value
+        elif not read_only and value_obtained_from_snapshot is not None:
+            os_sku = value_obtained_from_snapshot
+        else:
+            os_sku = raw_value
+
+        # this parameter does not need validation
+        return os_sku
+
+    def get_os_sku(self) -> Union[str, None]:
+        """Obtain the value of os_sku.
+
+        Note: Inherited and extended in aks-preview to add support for getting values from snapshot.
+
+        :return: string or None
+        """
+        return self._get_os_sku()
+
+    # pylint: disable=unused-argument
+    def _get_node_vm_size(self, read_only: bool = False, **kwargs) -> str:
+        """Internal function to dynamically obtain the value of node_vm_size according to the context.
+
+        If snapshot_id is specified, dynamic completion will be triggerd, and will try to get the corresponding value
+        from the Snapshot. When determining the value of the parameter, obtaining from `mc` takes precedence over user's
+        explicit input over snapshot over default vaule.
+
+        :return: string
+        """
+        default_value = "Standard_DS2_v2"
```

Comment:
Does 'az aks nodepool add' also use this _get_node_vm_size()? if so, I remembered that windows/linux nodepool has different default vm_size in old code.

Reply:
oh, this PR is for 'aks create' only, so the vm_size default value is good based on old aks_create code.